docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -182,118 +500,124 @@ pattern-propagators:
-- patterns:
-  - pattern: |
-      check_if_safe($X)
-      ...
-  - pattern: $X
+options:
+  taint_focus_on: source
 ```
 
-This has the same limitations as documented in [sources by side effect](#sources-by-side-effect). It is now recommended to set `by-side-effect: true` for the sanitizer in question. Enabling `by-side-effect` for a given sanitizer instructs Semgrep that whenever that sanitizer matches a variable (in general, an l-lvalue) exactly, then sanitization happens by side-effect. 
+<iframe src="https://semgrep.dev/embed/editor?snippet=JDPGP" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-## Sinks
+The [deduplication of findings](#deduplication-of-findings) still applies in this case. While Semgrep will now report all the taint sources, if a taint source can reach multiple sinks, the taint trace will only inform you about one of them.
 
-A sink is specified by a pattern but, unlike for sources and sanitizers, Semgrep finds the best match for the sink.
+## Minimizing false positives
 
-### Function argument as a sink
+The following [rule options](/writing-rules/rule-syntax/#options) can be used to minimize false positives:
 
-If you specify a sink such as `sink(...)` then any tainted data passed to sink, through any of its arguments, results in a match. You can narrow it down to a specific parameter this way: 
+| Rule option                   | Default | Description                                                            |
+| :-----------------------------| :------ | :--------------------------------------------------------------------- |
+| `taint_assume_safe_booleans`  | `false` | Boolean data is never considered tainted (works better with type annotations). |
+| `taint_assume_safe_numbers`   | `false` | Numbers (integers, floats) are never considered tainted (works better with type annotations). |
+| `taint_assume_safe_indexes`   | `false` | An index expression `I` tainted does not make an access expression `E[I]` tainted (it is only tainted if `E` is tainted). |
+| `taint_assume_safe_functions` | `false` | A function call like `F(E)` is not considered tainted even if `E` is tainted. (When using Pro's [inter-procedural taint analysis](#inter-procedural-analysis-pro), this only applies to functions for which Semgrep cannot find a definition.)  |
+| `taint_only_propagate_through_assignments` 🧪 | `false` | Disables all implicit taint propagation except for assignments. |
 
-```javascript
-pattern-sinks:
-  - patterns:
-    - pattern-inside: $S = new Sandbox(); ...
-    - pattern: $S.run($SINK, ...)
-    - focus-metavariable: $SINK
-```
+### Restrict taint by type (Pro)
 
-This rule causes Semgrep to only annotate the first parameter passed to `$S.run` as the sink, rather than the method `$S.run` itself. If taint goes into any other parameter of `$S.run`, then that is not considered a problem.
+By enabling `taint_assume_safe_booleans` Semgrep will automatically sanitize Boolean expressions (when it can infer that).
```

Comment:
```suggestion
By enabling `taint_assume_safe_booleans` Semgrep automatically sanitizes Boolean expressions when it can infer that the expression resolves to Boolean.
```

Reply:
Correct me if I'm mistaken, but the ability to infer is one of those things that Semgrep does at different levels depending on the language. For example, we do this for Java's stdlib, but I'm not aware of if we do this for other stdlibs?

Reply:
Stylistically, the reason why I replaced "when it can infer that" with a much longer phrase is because "that" (pronoun) can be a little vague. I just wanted to make the text clearer.